libs/hwes/eventstoredb/projections/custom/custom.go

```diff
@@ -0,0 +1,184 @@
+	if _, found := p.eventHandlers[eventType]; found {
+		zlog.Error().
+			Str("eventType", eventType).
+			Msg("override existing event handler")
+	}
+
+	p.eventHandlers[eventType] = eventHandler
+	return p
+}
+
+func (p *CustomProjection) handleEvent(ctx context.Context, event hwes.Event) (error, esdb.Nack_Action) {
+	eventHandler, found := p.eventHandlers[event.EventType]
+	if !found {
+		return fmt.Errorf("event type '%s' is invalid", event.EventType), esdb.Nack_Unknown
+	}
+	return eventHandler(ctx, event)
+}
+
+// Subscribe creates and subscribes to a persistent subscription in EventStoreDB
+// A persistent subscription is a type of subscription where the state is saved on the server-side
+// This function blocks the thread until the passed context gets canceled
+// https://developers.eventstore.com/server/v23.10/persistent-subscriptions.html
+func (p *CustomProjection) Subscribe(ctx context.Context) error {
+	log := zlog.Ctx(ctx)
+
+	// Create subscription on EventStoreDB
+	persistentAllSubscriptionOptions := esdb.PersistentAllSubscriptionOptions{}
+	// TODO: Do we need to manage the subscriptions? E.g. delete persistent subscriptions?
+	err := p.es.CreatePersistentSubscriptionAll(ctx, subscriptionGroupName, persistentAllSubscriptionOptions)
+	if err != nil {
+		// Ignore subscription already exists error
+		// The state of a persistent subscriptions is managed on the server-side by EventStoreDB.
+		// A persistent subscription must be created before connection.
+		// We ignore a failed creation to ensure idempotency.
+		if persistentSubscriptionError, ok := err.(*esdb.PersistentSubscriptionError); !ok || ok && (persistentSubscriptionError.Code != PersistentSubscriptionFailedCreationErrorCode) {
+			return err
+		} else {
+			log.Debug().Err(err).Msg("ignoring subscription already exists error")
+		}
+	}
+
+	// After a potential successful creation of a persistent subscription, we are trying to establish a connection to that subscription
+	stream, err := p.es.ConnectToPersistentSubscriptionToAll(ctx, subscriptionGroupName, esdb.ConnectToPersistentSubscriptionOptions{})
+	if err != nil {
+		return err
+	}
+	defer stream.Close()
+
+	log.Info().Msg("Start custom projection")
+	err = p.processReceivedEventsFromStream(ctx, stream)
+	log.Info().Msg("Stop custom projection")
+	return err
+}
+
+// processReceivedEventsFromStream acts as a worker on the subscribed stream
+// This functions tries to receive an event from the passed stream
+// and calls the according event handler based on the received event
+// This function blocks the thread until the passed context gets canceled
+func (p *CustomProjection) processReceivedEventsFromStream(ctx context.Context, stream *esdb.PersistentSubscription) error {
+	log := zlog.Ctx(ctx)
+
+	for {
+		esdbEvent := stream.Recv()
+
+		select {
+		case <-ctx.Done():
+			// Return when context gets finished
+			return nil
+		default:
+		}
+
```

Comment:
this is a good point to start a new span

Reply:
Yes. Required a refactor. Done.